### PR TITLE
Create explicit compiler C API libraries and enable dynamic linking on all platforms

### PR DESCRIPTION
**Pull request by stellaraccident**
This is the first step of a multi-part sequence that will:

* Rebuild the compiler/API directory into its final form.
* Move the compiler python API to its correct location (bindings/python).
* Enable dynamic linking project wide of all compiler tools.

This first step leaves the following undone:

* Python API stays where it is (there are a lot of fiddly path dependencies there and I opted to handle that in isolation).
* Sets everything up in an API2 which will eventually be renamed to API.
* Just gets the stub of what is needed for a delay loaded embedding API -- the full thing will follow and be integrated into PJRT for Jax and TensorFlow serving on the OSS side.
* The project local lld is still building static, even though the API includes it in the shared library. Was just trying to keep the patch size down vs there being a fundamental problem.

In the new state, there is a libIREECompiler.so.0 (or IREECompiler.dll on Windows) which contains both the in-process APIs and the tool entry-points. The following tools have been reworked to link against the shared library:

* iree-compile
* iree-lld (Python side only)
* iree-opt
* iree-mlir-lsp-server

In addition, some API tests which were doing expensive static linking now link dynamically. All said, this saves ~4-5x multi-gigabyte compiler tool linking steps and should be a pretty substantial build iteration improvement. I wasn't originally going to include iree-opt and iree-mlir-lsp-server in the shared build, but based on an experiment, they contribute negligible size (really just a few extra source files when linked together), and the productivity and usability savings (i.e. we can now distribute those at negligible cost) made me decide it was worth it.

I don't know what to do about iree-run-mlir. I think I can refactor it into something based on the compiler API that then links the runtime. But that is for later.

I took some steps to ensure this will work on Windows and MacOS but haven't finished verifying.

### CODE REVIEW

**Review by benvanik — APPROVED** [review]
This looks great to me! May want to get one more signoff from someone who may be responsible for integrating this into google3 - though because this is largely forked in API2 I suspect things may not break there immediately and there's leeway for iteration.

**Comment by stellaraccident** [comment]
> This looks great to me! May want to get one more signoff from someone who may be responsible for integrating this into google3 - though because this is largely forked in API2 I suspect things may not break there immediately and there's leeway for iteration.

I am reasonably confident that this improves the google3 integration side as more of the magic bits that were only done there and were done differently on the OSS/CMake side are now unified and guaranteed working at head. Bazel is still static linking tools, and it will continue to do so because it lacks any real facility for managing cross platform dynamic/runtime stuff (and the bug has been open for years without update) -- I have no plans to change Bazel to dynamic linking tools. That would be the big thing that would make it hard to support. I will be slowly extending it in follow-ons to support delay-loading a shared library plugin on that side, which will have some touch points internally but should be in the "new stuff" category enough that it isn't disruptive and can be fixed forward. I ran the basic idea by Geoffrey...

**Comment by benvanik** [comment]
image

**Comment by bjacob** [comment]
There is a postsubmit Windows test failure on my latest merged PR: https://github.com/iree-org/iree/actions/runs/3549048343 .

The log is unhelpful (AFAICS), just says 

```
The following tests FAILED:
    648 - iree/compiler/API2/test/api-test-binary (Exit code 0xc0000135
)
```

`Exit code 0xc0000135` is `STATUS_DLL_NOT_FOUND`

I wonder if the root cause might be this PR, but it's hard to tell, because the Windows postsubmit is skipped on this PR, and even when I hit "re-run all jobs" it skips it again:
https://github.com/iree-org/iree/actions/runs/3537484199

**Comment by GMNGeoffrey** [comment]
The full Windows build is currently postsubmit only:

https://github.com/iree-org/iree/blob/b35c97f3c23e1d85554f1dea50a0ef740c22c934/.github/workflows/ci.yml#L157

@ScottTodd do you think we could promote it to presubmit now?

**Comment by stellaraccident** [comment]
As noted on discord, should be a simple fix and I'll fix it forward. Glad we have *any* CI for it, frankly :)

**Comment by ScottTodd** [comment]
> @ScottTodd do you think we could promote it to presubmit now?

Yep. I'd still like to see if we can improve the repository and submodule checkout speed though... 6-10 minutes just to get the code is pretty slow :(

**Comment by stellaraccident** [comment]
> > @ScottTodd do you think we could promote it to presubmit now?
> 
> Yep. I'd still like to see if we can improve the repository and submodule checkout speed though... 6-10 minutes just to get the code is pretty slow :(

We can just tack it onto my 20 year old bug "Can you please, please replace NTFS with a modern file system"

**Comment by stellaraccident** [comment]
Fixed Windows: https://github.com/iree-org/iree/pull/11298

File: build_tools/bazel_to_cmake/bazel_to_cmake_targets.py [hunk 1]
```diff
@@ -125,7 +131,14 @@ def _convert_mlir_target(target):
   # Take "MLIR" and append the name part of the full target identifier, e.g.
   #   "@llvm-project//mlir:IR"   -> "MLIRIR"
   #   "@llvm-project//mlir:Pass" -> "MLIRPass"
-  return ["MLIR" + target.rsplit(":")[-1]]
+  # MLIR does not have header-only targets apart from the libraries. Here
```
Comment (GMNGeoffrey) [inline build_tools/bazel_to_cmake/bazel_to_cmake_targets.py:134]: Can we fix this upstream?

File: build_tools/bazel_to_cmake/bazel_to_cmake_converter.py [hunk 2]
```diff
@@ -371,6 +374,9 @@ def cc_test(self,
   def iree_runtime_cc_test(self, deps=[], **kwargs):
     self.cc_test(deps=deps + ["//runtime/src:runtime_defines"], **kwargs)
 
+  def iree_compiler_cc_test(self, deps=[], **kwargs):
+    self.cc_test(deps=deps + ["//compiler/src:defs"], **kwargs)
```
Comment (GMNGeoffrey) [inline build_tools/bazel_to_cmake/bazel_to_cmake_converter.py:378]: Hmmm I think we should be encoding this in the CMake, not the Bazel->CMake (same for the runtime one above)

File: build_tools/bazel_to_cmake/bazel_to_cmake_targets.py [hunk 3]
```diff
@@ -34,10 +32,17 @@
     "@llvm-project//llvm:X86AsmParser": ["IREELLVMCPUTargetDeps"],
     "@llvm-project//llvm:X86CodeGen": ["IREELLVMCPUTargetDeps"],
 
+    # LLD
+    "@llvm-project//lld": ["${IREE_LLD_TARGET}"],
+    "@llvm-project//lld:COFF": ["lldCOFF"],
+    "@llvm-project//lld:Common": ["lldCommon"],
+    "@llvm-project//lld:ELF": ["lldELF"],
+    "@llvm-project//lld:MachO": ["lldMachO"],
+    "@llvm-project//lld:Wasm": ["lldWasm"],
```
Comment (GMNGeoffrey) [inline build_tools/bazel_to_cmake/bazel_to_cmake_targets.py:41]: These look pretty standardized. We should handle it as a function with only lld special cased here (I'm still a bit suspicious of the lld special case also).

File: build_tools/bazel_to_cmake/bazel_to_cmake_targets.py [hunk 4]
```diff
@@ -51,6 +56,7 @@
     "@llvm-project//mlir:ShapeTransforms": ["MLIRShapeOpsTransforms"],
     "@llvm-project//mlir:ToLLVMIRTranslation": ["MLIRTargetLLVMIRExport"],
     "@llvm-project//mlir:mlir-translate": ["mlir-translate"],
+    "@llvm-project//mlir:MlirLspServerLib": ["MLIRLspServerLib"],
```
Comment (GMNGeoffrey) [inline build_tools/bazel_to_cmake/bazel_to_cmake_targets.py:59]: *sigh*. We should strip a leading "Mlir" from the Bazel targets in the conversion (and also fix this upstream)

File: compiler/src/iree/compiler/API/README.md [hunk 5]
```diff
@@ -1,11 +1,4 @@
...
-This is a top-level project for building public facing API packages that
-combine all dependent MLIR-based projects along with IREE's compiler backend
-API.
-
-It exports artifacts:
-
-* `iree-compiler-backend` Python wheel and source distributions, providing
-  the `iree.compiler_backend` Python packages.
-* Compiler C-API source and binary tarballs (future), with some CLI tools.
+This directory is being refactored into API2. Currently, it only contains
```
Comment (GMNGeoffrey) [inline compiler/src/iree/compiler/API/README.md:3]: I think we should probably still leave the documentation until it's actually deleted.

File: compiler/src/iree/compiler/API/python/CMakeLists.txt [hunk 6]
```diff
@@ -169,6 +162,43 @@ if(TARGET lld)
...
+# Install shared libraries that the extension depends on. This uses
+# CMake's defer feature to evaluate the install directive once everything
+# has been evaluated (because there is no guarantee that this file evaluates
+# before the API libraries are defined). While deferred calls are generally
+# fragile, this install is purely a leaf feature (with no other calls
+# depending on its sequencing), so this use is okay.
+# We defer it to the compiler/ directory so that cmake_install.cmake targets
+# are all self contained to the compiler, which is the most common spanning
+# parent.
+cmake_language(EVAL CODE "
```
Comment (GMNGeoffrey) [inline compiler/src/iree/compiler/API/python/CMakeLists.txt:174]: :scream:

File: compiler/src/iree/compiler/API/README.md (hunk 5 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/API/README.md:3]: > I think we should probably still leave the documentation until it's actually deleted.

Was my thought too. Until I read it and realized it was hopelessly wrong and better deleted.

File: compiler/src/iree/compiler/API/python/CMakeLists.txt (hunk 6 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/API/python/CMakeLists.txt:174]: Yeah. Right reaction. But I copied this from some code that I think Scott wrote, so I can't claim responsibility for this hack

**Pull request merged**